TableGenerator missing format width crash

```
> Task :violationReportConsole FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':violationReportConsole'.
> %-0s

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':violationReportConsole'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.LocalTaskInfoExecutor.execute(LocalTaskInfoExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:277)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:262)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:135)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.execute(DefaultTaskPlanExecutor.java:200)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:191)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.MissingFormatWidthException: %-0s
        at java.util.Formatter$FormatSpecifier.checkGeneral(Formatter.java:2954)
        at java.util.Formatter$FormatSpecifier.<init>(Formatter.java:2725)
        at java.util.Formatter.parse(Formatter.java:2560)
        at java.util.Formatter.format(Formatter.java:2501)
        at java.util.Formatter.format(Formatter.java:2455)
        at java.lang.String.format(String.java:2940)
        at net.twisterrob.gradle.quality.report.TableGenerator.build(TableGenerator.kt:49)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTaskKt.defaultAction(ValidateViolationsTask.kt:116)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTaskKt.access$defaultAction(ValidateViolationsTask.kt:1)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTask$action$1.invoke(ValidateViolationsTask.kt:27)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTask$action$1.invoke(ValidateViolationsTask.kt:24)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTaskKt$sam$org_gradle_api_Action$0.execute(ValidateViolationsTask.kt)
        at net.twisterrob.gradle.quality.tasks.ValidateViolationsTask.validateViolations(ValidateViolationsTask.kt:99)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:801)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:768)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        ... 34 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@
  * New model for Violations, a grouped property approach
  * New task for HTML report with limited Lint support
  * Tasks are now automatically added when applying `quality` plugin
+ * Fixed empty data handling in TableGenerator (#16)
 
 ## 0.6 *(2018-08-17 --- 2018-10-01)*
  * Gradle: 4.5.1

--- a/quality/src/main/kotlin/net/twisterrob/gradle/quality/report/TableGenerator.kt
+++ b/quality/src/main/kotlin/net/twisterrob/gradle/quality/report/TableGenerator.kt
@@ -37,15 +37,15 @@ class TableGenerator(
 		if (!printEmptyColumns) {
 			parsers = parsers.filter { summary[it] != null }
 		}
-		val format = parsers.map { Math.max(minWidth, it.length) }.joinToString(columnSeparator) { "%${it}s" }
+		val format = parsers.map { Math.max(minWidth, it.length) }.joinToString("") { "${columnSeparator}%${it}s" }
 		val longestModule = modules.maxBy { it.length }
 		val longestVariant = variants.maxBy { it.length }
 		val moduleWidth = Math.max(MIN_MODULE_LENGTH, longestModule?.length ?: 0)
 		val total = summary.values.sumBy { it ?: 0 }
-		val totalCountWidth = Math.log10(total.toDouble()).toInt()
+		val totalCountWidth = if (total == 0) 1 else Math.log10(total.toDouble()).toInt()
 		val variantWidth = Math.max(MIN_VARIANT_LENGTH + totalCountWidth, longestVariant?.length ?: 0)
 		val rowHeaderFormat = "%-${moduleWidth}s${columnSeparator}%-${variantWidth}s"
-		val rowFormat = "${rowHeaderFormat}${columnSeparator}${format}"
+		val rowFormat = "${rowHeaderFormat}${format}"
 		val header = String.format(rowFormat, *(listOf("module", "variant") + parsers).toTypedArray())
 		val rows = byModuleByVariantByParserCounts.flatMap { byModule ->
 			byModule.value.flatMap row@ { byVariant ->
@@ -65,7 +65,7 @@ class TableGenerator(
 			}
 		}
 		val footer = if (summaryRow) {
-			val summaryHeader = listOf("Summary", "(total ${total})")
+			val summaryHeader = listOf("Summary", "(total: ${total})")
 			val summaryData = parsers.map { summary[it]?.toString() ?: missingCount }
 			val summaryRow = String.format(rowFormat, *(summaryHeader + summaryData).toTypedArray())
 			listOf(summaryRow)

--- a/quality/src/test/kotlin/net/twisterrob/gradle/quality/report/TableGeneratorTest.kt
+++ b/quality/src/test/kotlin/net/twisterrob/gradle/quality/report/TableGeneratorTest.kt
@@ -43,7 +43,7 @@ class TableGeneratorTest {
 			module2	variant2 	       N/A	N/A	      N/A
 			module2	variant3 	         0	N/A	      N/A
 			module3	variant2 	       N/A	N/A	      N/A
-			Summary	(total 6)	         1	  5	      N/A
+			Summary	(total: 6)	         1	  5	      N/A
 		""".prepare()
 		assertEquals(expected, result)
 	}
@@ -58,7 +58,7 @@ class TableGeneratorTest {
 			module1	variant1 	         1	  2	      N/A
 			module1	variant2 	       N/A	  3	      N/A
 			module2	variant3 	         0	N/A	      N/A
-			Summary	(total 6)	         1	  5	      N/A
+			Summary	(total: 6)	         1	  5	      N/A
 		""".prepare()
 		assertEquals(expected, result)
 	}
@@ -75,7 +75,7 @@ class TableGeneratorTest {
 			module2	variant2 	       N/A	N/A
 			module2	variant3 	         0	N/A
 			module3	variant2 	       N/A	N/A
-			Summary	(total 6)	         1	  5
+			Summary	(total: 6)	         1	  5
 		""".prepare()
 		assertEquals(expected, result)
 	}
@@ -90,7 +90,58 @@ class TableGeneratorTest {
 			module1	variant1 	         1	  2
 			module1	variant2 	       N/A	  3
 			module2	variant3 	         0	N/A
-			Summary	(total 6)	         1	  5
+			Summary	(total: 6)	         1	  5
+		""".prepare()
+		assertEquals(expected, result)
+	}
+
+	@Test fun `empty data (no modules)`() {
+		val sut = TableGenerator()
+
+		val result = sut.build(emptyMap())
+
+		val expected = """
+			module 	variant${"   "}
+			Summary	(total: 0)
+		""".prepare()
+		assertEquals(expected, result)
+	}
+
+	@Test fun `empty data (no variants)`() {
+		val sut = TableGenerator()
+
+		val result = sut.build(
+			mapOf(
+				"module1" to emptyMap(),
+				"module2" to emptyMap(),
+				"module3" to emptyMap()
+			)
+		)
+
+		val expected = """
+			module 	variant${"   "}
+			Summary	(total: 0)
+		""".prepare()
+		assertEquals(expected, result)
+	}
+
+	@Test fun `empty data (no parsers)`() {
+		val sut = TableGenerator()
+
+		val result = sut.build(
+			mapOf(
+				"module1" to mapOf("variant1" to emptyMap()),
+				"module2" to mapOf("variant2" to emptyMap()),
+				"module3" to mapOf("variant3" to emptyMap())
+			)
+		)
+
+		val expected = """
+			module 	variant${"   "}
+			module1	variant1${"  "}
+			module2	variant2${"  "}
+			module3	variant3${"  "}
+			Summary	(total: 0)
 		""".prepare()
 		assertEquals(expected, result)
 	}

